core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -265,6 +277,68 @@ private boolean isWriteSetOverlappedWith(Scan scan) {
     return false;
   }
 
+  private boolean isWriteSetOverlappedWithRelational(Scan scan) {
+    for (Map.Entry<Key, Put> entry : writeSet.entrySet()) {
+      if (scanSet.get(scan).contains(entry.getKey())) {
+        return true;
+      }
+
+      // the following check prevents scan-after-write even if the update record potentially matches
```

Comment:
Could you elaborate this? It would be great if you give me an example problematic case.

Reply:
We need to consider three cases here to prevent scan-after-write.
1) A put operation overlaps the scan range regardless of the update (put) results.
2) A put operation does not overlap the scan range as a result of the update.
3) A put operation overlaps the scan range as a result of the update.

We can check cases 1 and 2 with L. 282-284. (Note that case 2 does not have the overlap, but we intentionally prohibit it due to the consistency and simplicity of snapshot management.) For case 3, we need to evaluate the operator to check for overlap since the put operation is only in the read-write set and not committed to the database. Let's see the following examples.

```sql
-- Assume the following table.
-- tbl (key int, val int)
-- 1 | 2 
-- 2 | 3

-- Case 2
update tbl set val = 4 where key = 1;
select * from tbl where val < 3;
-- We can find the overlap using the scan results.

-- Case 3
update tbl set val = 2 where key = 2;
select * from tbl where val < 3;
-- We cannot find the overlap since the database is not updated yet.
```

Reply:
Thanks a lot for the explanation!!

Reply:
@jnmt This is just a suggestion and doesn't prevent this PR from being merged. The above explanation was really helpful for me to understand why we need to consider updated values. How about putting it as a comment so that new developers and 2y later me who has a bad memory can maintain the code?

Reply:
Yeah, I guess I will also forget it. It's a bit too much, but I added the detailed comments in 4b2fbe5.

Reply:
💯